lib/ClangImporter/ClangClassTemplateNamePrinter.cpp

```diff
@@ -59,9 +59,37 @@ struct TemplateInstantiationNamePrinter
         namedArg = typeDefDecl;
       llvm::SmallString<128> storage;
       llvm::raw_svector_ostream buffer(storage);
-      nameImporter->importName(namedArg, version, clang::DeclarationName())
-          .getDeclName()
-          .print(buffer);
+
+      // Print the fully-qualified type name.
+      std::stack<DeclName> qualifiedNameComponents;
```

Comment:
Is it sufficient to only do this for tag types? Not sure how type aliases are printed, but I could imagine getting into similar situations with those.

Reply:
Hmm, interesting. Typedefs are already expanded into their underlying types at this point, e.g.
```cpp
using A = B<C>;
using D = E<A>; // will be named `struct E<B<C>>` 
```
Do you think this would be problematic in some cases?

Reply:
Oh, I see. I don't anticipate problems due to the desugaring, I am just wondering if this Is a minor usability issue in case these type names are ever surfaced to the users. Anyhow, I think if we want to change this it belongs to a separate PR.

Reply:
After some more thought, desugaring type aliases is the right thing to do for anything mangling related and this sidesteps all the issues I had in mind so I have no concerns here.